node/core/backing/src/lib.rs

```diff
@@ -0,0 +1,1454 @@
+		}
+
+		Ok(())
+	}
+
+	async fn import_statement(&mut self, statement: SignedFullStatement) -> Result<(), Error> {
+		let idx = statement.validator_index() as usize;
+
+		if self.table_context.validators.len() > idx {
+			match statement.check_signature(
+				&self.table_context.signing_context,
+				&self.table_context.validators[idx],
+			) {
+				Ok(()) => {
+					let stmt = primitive_statement_to_table(&statement);
+
+					self.table.import_statement(&self.table_context, stmt);
+
+					self.issue_new_misbehaviors().await?;
+				}
+				Err(()) => {
+					return Err(Error::ValidationFailed(ValidationFailed));
+				}
+			}
+		};
+
+		Ok(())
+	}
+
+	async fn process_msg(&mut self, msg: CandidateBackingMessage) -> Result<(), Error> {
+		match msg {
+			CandidateBackingMessage::Second(_, candidate, pov) => {
+				// If the message is a `CandidateBackingMessage::Second`, sign and dispatch a
+				// Seconded statement only if we have not seconded any other candidate and
+				// have not signed a Valid statement for the requested candidate.
+				match self.seconded {
+					// This job has not seconded a candidate yet.
+					None => {
+						let candidate_hash = candidate.hash();
+
+						if !self.issued_validity.contains(&candidate_hash) {
+							if let Ok(ValidationResult::Valid) = self.validate_and_second(
+								candidate,
+								pov,
+							).await {
+								self.seconded = Some(candidate_hash);
+							}
+						}
+					}
+					// This job has already seconded a candidate.
+					Some(_) => {}
+				}
+			}
+			CandidateBackingMessage::Statement(_, statement) => {
+				self.maybe_validate_and_import(statement).await?;
+			}
+			CandidateBackingMessage::GetBackedCandidates(_, tx) => {
+				let backed = self.get_backed();
+
+				tx.send(backed).map_err(|_| oneshot::Canceled)?;
+			}
+		}
+
+		Ok(())
+	}
+
+	async fn maybe_validate_and_import(
+		&mut self,
+		statement: SignedFullStatement,
+	) -> Result<(), Error> {
+		let idx = statement.validator_index() as usize;
+
+		if self.table_context.validators.len() > idx {
+			if let Err(()) = statement.check_signature(
+				&self.table_context.signing_context,
+				&self.table_context.validators[idx],
+			) {
+				return Err(Error::InvalidSignature);
+			}
+		} else {
+			return Err(Error::InvalidSignature);
+		}
+
+		if let Statement::Seconded(candidate) = statement.payload() {
+			if candidate.parachain_index == self.assignment {
+				let pov = self.request_pov_from_distribution(candidate.clone().into()).await?;
+				let v = self.request_candidate_validation(candidate.clone(), pov).await?;
+
+				let statement = match v {
+					ValidationResult::Valid => Statement::Valid(candidate.hash()),
+					ValidationResult::Invalid => Statement::Invalid(candidate.hash()),
+				};
+
+				if let Some(signed_statement) = self.sign_statement(statement) {
+					self.distribute_signed_statement(signed_statement).await?;
+				}
+			}
+		}
+
+		self.import_statement(statement).await?;
```

Comment:
`import_statement` also does the signature check. IMO it should be done as early as possible and only once.